### PR TITLE
Check redactions on restoration

### PR DESCRIPTION
**Pull request by bethanyaconnor**
[Tech spec](https://docs.google.com/document/u/1/d/1SHKHYmcITP7N7sDhWZpyUI7nPR8xkwIvBisSCwOG5kY/edit)

This PR adds a flag to restoration to clear any strings that don't have exactly the expected set of redactions. This is done by a post-restoration check, which might cause some performance hits (I wouldn't expect it to be too bad).

### CODE REVIEW

File: src/bin/restore.js [hunk 1]
```diff
@@ -42,8 +43,19 @@ if (compilerPlugins) {
   processor.compilerPlugins.push(...requireByPath(compilerPlugins));
 }
 
+var check = false
+if (argv.v || argv.checkRestorations) {
+  check = true;
+}
+
+function restore_with_check() {
```
Comment (Hamms) [inline src/bin/restore.js]: why is this necessary? Can't we just do `recursivelyProcessAll(processor.sourceAndRedactedToRestored.bind(processor, check), data);`?

File: src/bin/restore.js [hunk 2]
```diff
@@ -21,6 +21,7 @@ if (helpFlag || missingRequiredFlags) {
   process.stdout.write("\t-f, --format [md|txt]: specify format of content (default to markdown)\n");
   process.stdout.write("\t-p, --parserPlugins PLUGINS: comma-separated list of parser plugins to include in addition to the defaults\n");
   process.stdout.write("\t-c, --compilerPlugins PLUGINS: comma-separated list of compiler plugins to include in addition to the defaults\n");
+  process.stdout.write("\t-v, --checkRestorations Discard restoration if redactions are added or missing");
```
Comment (Hamms) [inline src/bin/restore.js]: why `v`?

File: src/plugins/process/restoreRedactions.js [hunk 3]
```diff
@@ -93,9 +101,18 @@ module.exports = function restoreRedactions(sourceTree) {
...
+            content: content
+          });
+        }
         return;
       }
-
+      if (check && redactedData.used) {
+        return unrestored(eat(match[0]), redactedData);
+      }
+      redactedData.used = true;
```
Comment (Hamms) [inline src/plugins/process/restoreRedactions.js]: this seems unnecessarily complicated. What about something like

```js
      const redactedData = redactions[index];
      if (!redactedData) {
        if (check) {
          return eat(match[0])({
            type: 'unrestored',
            content: content
          });
        }
        return;
      }
      redactions[index] = undefined;
```

File: src/plugins/process/restoreRedactions.js [hunk 4]
```diff
@@ -106,7 +123,9 @@ module.exports = function restoreRedactions(sourceTree) {
       }
 
       const add = eat(match[0]);
-      return restorationMethod(add, redactedData, content);
+      var node = restorationMethod(add, redactedData, content);
```
Comment (Hamms) [inline src/plugins/process/restoreRedactions.js]: nit (applies here and elsewhere): prefer `const` or `let` over `var`

File: src/redactableProcessor.js [hunk 5]
```diff
@@ -43,23 +45,61 @@ module.exports = class RedactableProcessor {
...
 
-  sourceAndRedactedToMergedSyntaxTree(source, redacted) {
-    const sourceTree = this.sourceToRedactedSyntaxTree(source);
-    const redactedTree = this.getProcessor()
+  /* This function checks that
+   * 1. Redactions in the sourceTree == restorations performed
+   * 2. There are no redactions that weren't performed
+   * 3. There was no additions redactable content in the mergedTree
+   * */
+  checkRestorationNodes(sourceTree, mergedTree) {
```
Comment (Hamms) [inline src/redactableProcessor.js]: is it necessary to check all these things? Seems like we could just check for `redaction` and `unrestored` nodes in the mergedTree, which would have the added benefit of not adding the `restored` property to MDAST nodes

File: src/redactableProcessor.js [hunk 6]
```diff
@@ -43,23 +45,61 @@ module.exports = class RedactableProcessor {
...
+      }
+    });
+
+    return source_redactions === redacted_restorations &&
+           redacted_redactions === 0 &&
+           redacted_unrestored === 0;
+  }
+
+  sourceAndRedactedToMergedSyntaxTree(sourceTree, redacted, check) {
+    var settings = { redact: check, check: check};
```
Comment (Hamms) [inline src/redactableProcessor.js]: why `redact: check`?

File: src/redactableProcessor.js [hunk 7]
```diff
@@ -43,23 +45,61 @@ module.exports = class RedactableProcessor {
...
+    const sourceTree = this.sourceToRedactedSyntaxTree(source);
     const mergedSyntaxTree = this.sourceAndRedactedToMergedSyntaxTree(
-      source,
-      redacted
+      sourceTree,
+      redacted,
+      check
     );
+    if (check && !this.checkRestorationNodes(sourceTree, mergedSyntaxTree)) {
+      return "";
```
Comment (Hamms) [inline src/redactableProcessor.js]: this should probably be `"\n"`, for consistency

File: src/redactableProcessor.js [hunk 8]
```diff
@@ -43,23 +45,61 @@ module.exports = class RedactableProcessor {
...
+    }
     return this.getProcessor()
       .use(this.constructor.getCompiler())
       .use(this.compilerPlugins)
+      .use(renderRedactions)
```
Comment (Hamms) [inline src/redactableProcessor.js]: I'd like to see a comment explaining why this is here

File: test/integration/data/malformed-redactions-json/restored.json [hunk 9]
```diff
@@ -0,0 +1,12 @@
+{
+  "basic example": {
```
Comment (Hamms) [inline test/integration/data/malformed-redactions-json/restored.json]: for these examples, could you instead do something like, `malformed redaction`, `missing redaction`, `extra redaction`, `content that should have been redacted`, etc?

File: test/integration/data/malformed-redactions-md/translated.md [hunk 10]
```diff
@@ -0,0 +1 @@
+This is an example of a redacted [link][0] and an [added one](extra.com)
```
Comment (Hamms) [inline test/integration/data/malformed-redactions-md/translated.md]: :+1:

File: test/integration/scripts.test.js [hunk 11]
```diff
@@ -80,7 +80,7 @@ describe("Command-Line Scripts", () => {
             }
             const restore = spawnSync('node', args);
             let expectedContent = fs.readFileSync(expected, 'utf8')
-            if (extension === extensions.text) {
+            if (extension !== extensions.markdown) {
```
Comment (Hamms) [inline test/integration/scripts.test.js]: is there a specific reason for this switch?

File: test/unit/redactableMarkdownProcessor.test.js [hunk 12]
```diff
@@ -111,5 +111,15 @@ describe('Standard Markdown', () => {
       const output = processor.sourceAndRedactedToHtml(source, redacted);
       expect(output).toEqual("<p>C'est du texte avec <a href=\"http://example.com/\">un lien</a> et [une image][1]</p>\n");
     });
+
+    it('will handle extra/unwanted redactions by treating them as references without definitions with flag set', () => {
```
Comment (Hamms) [inline test/unit/redactableMarkdownProcessor.test.js]: I'd love to see tests for the other failure cases!

File: test/integration/scripts.test.js (hunk 11 above)
Comment (bethanyaconnor) [reply test/integration/scripts.test.js]: I was having problems with added newlines to the expected content, despite there not being any in the files.

File: src/redactableProcessor.js (hunk 5 above)
Comment (bethanyaconnor) [reply src/redactableProcessor.js]: I think the check that we restore the right number of redactions is necessary to detect missing redactions. I can't think of any other way to check this except to keep track of `restored` nodes.

File: src/bin/restore.js [hunk 13]
```diff
@@ -42,8 +43,19 @@ if (compilerPlugins) {
...
+
+function restore_with_strict() {
+  return function(...args) {
+    return processor.sourceAndRedactedToRestored(...args, strict);
+  }
+}
+
 function restore(data) {
-  return recursivelyProcessAll(processor.sourceAndRedactedToRestored.bind(processor), data);
+  return recursivelyProcessAll(restore_with_strict().bind(processor), data);
```
Comment (Hamms) [inline src/bin/restore.js]: (following up with https://github.com/code-dot-org/redactable-markdown/pull/63#discussion_r302211861)

Oooh, I see. In that case, how about something more explicit, like

```js
return recursivelyProcessAll((source, redacted) => processor.sourceAndRedactedToRestored(source, redacted, strict), data);
```

File: src/plugins/process/restoreRedactions.js [hunk 14]
```diff
@@ -106,7 +115,9 @@ module.exports = function restoreRedactions(sourceTree) {
       }
 
       const add = eat(match[0]);
-      return restorationMethod(add, redactedData, content);
+      let node = restorationMethod(add, redactedData, content);
```
Comment (Hamms) [inline src/plugins/process/restoreRedactions.js]: nit:
```suggestion
      const node = restorationMethod(add, redactedData, content);
```

File: src/plugins/process/restoreRedactions.js [hunk 15]
```diff
@@ -188,7 +209,14 @@ module.exports = function restoreRedactions(sourceTree) {
       const subvalue = value.slice(startIndex, endIndex);
       const children = this.tokenizeBlock(subvalue, eat.now());
       const add = eat(blockOpen + subvalue + blockClose);
-      return restorationMethod(add, redactedData, content, children);
+      let nodes = restorationMethod(add, redactedData, content, children);
```
Comment (Hamms) [inline src/plugins/process/restoreRedactions.js]: nit:
```suggestion
      const nodes = restorationMethod(add, redactedData, content, children);
```

File: src/redactableProcessor.js (hunk 5 above)
Comment (Hamms) [reply src/redactableProcessor.js]: Ahh, of course.

I mostly bring this up because I really don't like adding a property to [MDAST nodes](https://github.com/syntax-tree/mdast). It violates a contract and introduces an element of state to the syntax tree that really shouldn't be there.

What if we include the redacted tree in this check? Something like

```js
const selectAll = require('unist-util-select').selectAll;
const select = require('unist-util-select').select

const sourceRedactions = selectAll('redaction', sourceTree);
const redactedRedactions = selectAll('redaction', redactedTree);

// If source and redacted have mismatching number of redactions, this
// means a redaction was either added or removed in the redacted text
if (sourceRedactions.length !== redactedRedactions.length) {
  return false;
}

// Even if both strings have the same total number of redactions, they
// may have mismatching redaction indexes or other errors that prevent
// perfect restoration
if (select('unrestored', mergedTree)) {
  return false;
}

return true;
```

Comment (Hamms) [reply src/redactableProcessor.js]: It occurs to me this would require us to be able to parse redacted text into a syntax tree, which we can't currently do. But I think that's a good thing! Right now, restoration happens as part of the parsing process, which is really strange; specifically, it's a thing that requires a syntax tree and some text, and produces a second syntax tree. What if we instead added the ability to parse redacted text to a syntax tree, and then made restoration be a process that takes two syntax trees and produces a third? That seems _much_ cleaner

File: test/integration/data/malformed-redactions-json/translated.json [hunk 16]
```diff
@@ -0,0 +1,11 @@
+{
+  "basic example": {
+    "good string": "This has a [link][0] that will stay put",
+    "good string with repeated link": "This has a [link][0] but the link [repeats][1]",
+    "added redactable content": "This has a [link][0] and will have redactable content added as another [link][1]",
+    "removed redactable content": "This has a [link][0]",
+    "reused redaction": "This has a [link][0] and will add a redaction with the same index [link number 2][0]",
+    "manufactured redaction": "This has a [link][0] and will add a redaction with a different index [made up link][1]",
+    "malformed redaction": "This [link] [0] redaction will have broken formatting"
```
Comment (Hamms) [inline test/integration/data/malformed-redactions-json/translated.json]: I'd also love to see something like

```js
source: "This has just one [link](example.com)",
redacted: "This has just one [link][0]",
translated: "This has the original [link][0] and also an [extra link](example.com)"
restored: ""
```

File: src/bin/restore.js (hunk 13 above)
Comment (bethanyaconnor) [reply src/bin/restore.js]: Oh that is bettter!

File: src/redactableProcessor.js [hunk 17]
```diff
@@ -70,13 +108,23 @@ module.exports = class RedactableProcessor {
...
 
-  sourceAndRedactedToRestored(source, redacted) {
+  sourceAndRedactedToRestored(source, redacted, strict) {
     const sourceTree = this.sourceToRedactedSyntaxTree(source);
-    const restorationTree = this.redactedToSyntaxTree(redacted);
+    const restorationTree = this.redactedToSyntaxTree(redacted, strict);
     const mergedSyntaxTree = this.sourceAndRedactedToMergedSyntaxTree(
       sourceTree,
-      restorationTree
+      JSON.parse(JSON.stringify(restorationTree))
```
Comment (Hamms) [inline src/redactableProcessor.js]: what's this for?

Comment (bethanyaconnor) [reply src/redactableProcessor.js]: The transformer modifies the tree, which means that the verification logic can't rely on information from the `restorationTree`.

I found this method of deep copying in a couple of place ([example](https://www.codementor.io/ramnmiklus/deep-copying-an-object-in-javascript-mdlj2c318#a-solution)) but definitely open to suggestions of a better way. I'll add a comment here at the very least.

Comment (Hamms) [reply src/redactableProcessor.js]: Oh, true. I don't love that that's a side effect. I'd vote we rewrite the transformer to use something like https://github.com/syntax-tree/unist-util-map to create a new tree. Alternatively, the "deep copy" logic should probably happen in the `sourceAndRedactedToMergedSyntaxTree` method itself

Comment (bethanyaconnor) [reply src/redactableProcessor.js]: Yeah I don't love the side effect either. But, I remember trying to use `map` instead of `visit` and it caused all sorts of problems with block restorations because we might change multiple blocks at once.

**Comment by bethanyaconnor** [comment]
Created https://codedotorg.atlassian.net/browse/FND-724 to track.
